### PR TITLE
test: ensure utf-8 sources/archives filenames work properly

### PR DESCRIPTION
**Pull request by caarlos0**
This now works for files added to the source archive **after** the `git archive` command is run. During the `git archive`, it seems that the `tar` is still using some format that doesn't support utf-8 by default.

Tomorrow I'll look more into it.

see https://pkg.go.dev/archive/tar#Format

closes #3812

### CODE REVIEW

**Comment by caarlos0** [comment]
@mohammed90 no idea tbh

**Comment by caarlos0** [comment]
I'm trying to come up with a reproducible for the Linux issue, so far no look...

if you wanna try it too: https://go.dev/play/p/57CO6WXxoPN

**Comment by caarlos0** [comment]
that said, forcing GNU headers seems to fix it... not sure if its the right fix though

**Comment by mohammed90** [comment]
I figured it out 🕵🏼‍♂️ 

It comes mangled from the output of `git ls-files`, because Git is weird: [[How to make Git properly display UTF-8 encoded pathnames in the console window?](https://stackoverflow.com/questions/22827239/how-to-make-git-properly-display-utf-8-encoded-pathnames-in-the-console-window)](https://stackoverflow.com/questions/22827239/how-to-make-git-properly-display-utf-8-encoded-pathnames-in-the-console-window).

The issue was git all along. Goreleaser used to call `git ls-files`, then parsing those lines to figure out which files to add, appending them to the list of files. The output coming from `git ls-files` was messed up due to the description detailed in the StackOverflow link. Part of your refactor removed the call to `git ls-files`, which side-stepped git shenanigans of printing non-ASCII characters in octal. The fix was a side-effect of removing that line, not due to tar headers. Why Linux? I don't know yet, but I'll chalk it up to Git being itself 💁🏼‍♂️ 

All in all, I think this change is still good because it expands the compatibility, but it's up to you. I can rest in peace now.

**Comment by caarlos0** [comment]
oh, awesome you find it out!

yeah, I was already convinced that the issue would either be git or somehow Go's tar package... I was digging into tar trying to find something, but everything seemed fine.

One thing that still bothers me (because I don't understand why it happens yet):

- I create the archive with `git archive`
- I scan the archive and add the same headers/files to another archive - and I checked, they are all in PAX format
- the files with emojis and etc, even though they are added to the new archive, are not present in the final result
- if I force the all headers to GNU, it works

so, I'll investigate that as well before merging, just to make sure its not hiding some other issuel...

**Comment by mohammed90** [comment]
I did a quick search and couldn't find anything meaningful. However, it seems like Gnu recommends moving to PAX format: https://www.gnu.org/software/tar/manual/html_section/Formats.html

**Comment by caarlos0** [comment]
okay, this last part was on me, I feel dumb even, but figured it out...

tests were failing because my method that lists files in a tar file was ignoring pax headers 👎🏻 

anyway, fixed that, now it works!

File: pkg/archive/tar/tar.go [hunk 1]
```diff
@@ -30,13 +30,15 @@ func Copying(source io.Reader, target io.Writer) (Archive, error) {
...
-		h, err := r.Next()
-		if err == io.EOF || h == nil {
+		header, err := r.Next()
+		if err == io.EOF || header == nil {
 			break
 		}
-
-		w.files[h.Name] = true
-		if err := w.tw.WriteHeader(h); err != nil {
+		if err != nil {
```
Comment (caarlos0) [inline pkg/archive/tar/tar.go:37]: missing error handling, only real change in this pr

File: internal/testlib/archive.go [hunk 2]
```diff
@@ -136,7 +136,7 @@ func doLsTar(f io.Reader) []string {
 		if h == nil || err == io.EOF {
 			break
 		}
-		if h.Format == tar.FormatPAX {
+		if h.Name == "pax_global_header" {
```
Comment (caarlos0) [inline internal/testlib/archive.go:139]: something git-archive does, apparently

**Pull request merged**